### PR TITLE
Add support for interprocess change notifications

### PR DESCRIPTION
**Pull request by tgoyne**

### CODE REVIEW

File: Realm/RLMRealm.mm [hunk 1]
```diff
@@ -196,16 +187,26 @@ static void clearRealmCache() {
         for (NSMapTable *map in s_realmsPerPath.allValues) {
             [map removeAllObjects];
         }
-        s_realmsPerPath = [NSMutableDictionary dictionary];
+        [s_realmsPerPath removeAllObjects];
     }
 }
 
+static NSMutableDictionary *s_listenersPerPath = [NSMutableDictionary new];
```
Comment (alazier) [inline Realm/RLMRealm.mm]: we should clear this in `resetRealmState`

File: Realm/RLMRealm.mm [hunk 2]
```diff
@@ -196,16 +187,26 @@ static void clearRealmCache() {
...
+// must be called with s_listenersPerPath locked
+static RLMChangeListener *listenerForPath(NSString *path, BOOL inMemory, BOOL create=YES) {
+    RLMChangeListener *listener = s_listenersPerPath[path];
```
Comment (alazier) [inline Realm/RLMRealm.mm]: synchronization should be pushed from the caller into this and other listener api methods

File: Realm/RLMRealm.mm [hunk 3]
```diff
@@ -196,16 +187,26 @@ static void clearRealmCache() {
...
             [map removeAllObjects];
         }
-        s_realmsPerPath = [NSMutableDictionary dictionary];
+        [s_realmsPerPath removeAllObjects];
     }
 }
 
+static NSMutableDictionary *s_listenersPerPath = [NSMutableDictionary new];
+// must be called with s_listenersPerPath locked
+static RLMChangeListener *listenerForPath(NSString *path, BOOL inMemory, BOOL create=YES) {
```
Comment (alazier) [inline Realm/RLMRealm.mm]: Should split this into two methods `startListening(Realm *realm)` and `stopListening(Realm *realm)`

File: Realm/RLMChangeListener.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+// A thread which waits for change notifications on the given path and notifies
+// all registered RLMRealms when a change occurs. Does *not* retain registered
+// RLMRealm instances
+@interface RLMChangeListener : NSThread
+- (instancetype)initWithPath:(NSString *)path
+                    inMemory:(BOOL)inMemory
+                       cache:(NSMutableDictionary *)cache;
+
+// This both must be called with `cache` locked
```
Comment (segiddins) [inline Realm/RLMChangeListener.h]: s/This/These methods/

File: Realm/RLMChangeListener.mm [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+- (void)run {
+    while (_sharedGroup && _sharedGroup->wait_for_change()) {
+        // we don't have any accessors, so just start a new read transaction
+        // rather than using advance_read() as that does far more work
+        _sharedGroup->end_read();
+        _sharedGroup->begin_read();
+
+        @synchronized (_realms) {
```
Comment (alazier) [inline Realm/RLMChangeListener.mm]: would be nice to have an explicit check that we cancelled/stopped notifications for this path

File: Realm/RLMChangeListener.mm [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    std::unique_ptr<SharedGroup> _sharedGroup;
+    NSString *_path;
+    NSMutableArray *_realms;
+    NSCondition *_shutdownCondition;
+}
+
+- (instancetype)initWithPath:(NSString *)path inMemory:(BOOL)inMemory cache:(NSMutableDictionary *)cache {
+    self = [super initWithTarget:self selector:@selector(run) object:nil];
+    if (self) {
+        _cache = cache;
```
Comment (alazier) [inline Realm/RLMChangeListener.mm]: passing in a global cache to be used by all instances seems wrong

File: Realm/RLMChangeListener.mm [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        _replication.reset(tightdb::makeWriteLogCollector(path.UTF8String));
+        SharedGroup::DurabilityLevel durability = inMemory ? SharedGroup::durability_MemOnly :
```
Comment (alazier) [inline Realm/RLMChangeListener.mm]: Why not create/open the shared group in the background thread?

File: Realm/RLMChangeListener.h [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+////////////////////////////////////////////////////////////////////////////
+
+#import <Foundation/Foundation.h>
+
+@class RLMRealm;
+
+// A thread which waits for change notifications on the given path and notifies
+// all registered RLMRealms when a change occurs. Does *not* retain registered
+// RLMRealm instances
+@interface RLMChangeListener : NSThread
```
Comment (alazier) [inline Realm/RLMChangeListener.h]: Would be preferable to just use a block rather than subclassing NSThread

Comment (alazier) [inline Realm/RLMChangeListener.h]: Or at least use an NSThread instance instead of subclassing

File: Realm/RLMChangeListener.mm [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        _path = path;
+        _realms = [NSMutableArray array];
+        _shutdownCondition = [[NSCondition alloc] init];
+        [self start];
+    }
+    return self;
+}
+
+- (void)addRealm:(RLMRealm *)realm {
+    @synchronized (_realms) {
```
Comment (alazier) [inline Realm/RLMChangeListener.mm]: Would be good to only synchronize on _cache (or the global once it is moved into this file).

File: Realm/RLMChangeListener.mm [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        [_realms addObject:[[RLMWeakNotifier alloc] initWithRealm:realm]];
+    }
+}
+
+- (void)removeRealm:(RLMRealm *)realm {
+    @synchronized (_realms) {
+        @autoreleasepool {
+            // The NSPredicate needs to be deallocated before we return or it crashes,
+            // since we're called from -dealloc and so retaining `realm` doesn't work.
+            [_realms filterUsingPredicate:[NSPredicate predicateWithFormat:@"realm != nil AND realm != %@", realm]];
```
Comment (alazier) [inline Realm/RLMChangeListener.mm]: Use a NSMutableSet instead so you don't have to do this query for removal

**Comment by tgoyne** [comment]
From #1531:

> The remaining test failure is a perf regression for realm creation. Not a lot of room left for optimization since 80% of the time is spent in the three `open()` calls. One thing we could do is push the `open()` on the fifo to the background thread, but that'd limit how many error cases we could report gracefully.

**Comment by tgoyne** [comment]
retest me please

**Comment by tgoyne** [comment]
retest this please

File: Realm/RLMProperty.mm [hunk 11]
```diff
@@ -39,7 +39,6 @@ - (instancetype)initWithName:(NSString *)name
         _objectClassName = objectClassName;
         _indexed = indexed;
         [self setObjcCodeFromType];
-        [self updateAccessors];
```
Comment (alazier) [inline Realm/RLMProperty.mm]: Why do we no longer need this?

Comment (tgoyne) [inline Realm/RLMProperty.mm]: We never needed (there) it in the first place. That initializer is only used for dynamic objects, and dynamic objects don't have accessor properties.

Comment (tgoyne) [inline Realm/RLMProperty.mm]: That change can be split out of this PR if desired; it was just a perf optimization to try to make up for the added overhead of opening the named pipe.

Comment (alazier) [inline Realm/RLMProperty.mm]: Lets split it out for now.

File: Realm/RLMRealm.mm [hunk 12]
```diff
@@ -395,7 +335,7 @@ + (instancetype)realmWithPath:(NSString *)path
     }
 
     // try to reuse existing realm first
-    RLMRealm *realm = cachedRealm(path);
+    RLMRealm *realm = RLMGetCurrentThreadCachedRealmForPath(path);
```
Comment (alazier) [inline Realm/RLMRealm.mm]: This sounds a bit awkward. `RLMGetThreadLocalCachedRealmForPath` sounds a bit better. This isn't a big deal though.

File: Realm/RLMRealm.mm [hunk 13]
```diff
@@ -450,10 +391,17 @@ + (instancetype)realmWithPath:(NSString *)path
...
         }
+
+        if (!dynamic) {
+            RLMCacheRealm(realm);
+        }
     }
 
-    if (!dynamic) {
-        cacheRealm(realm, path);
+    if (!readonly) {
```
Comment (alazier) [inline Realm/RLMRealm.mm]: should this be in the synchronized block?

File: Realm/RLMRealm.mm (hunk 12 above)
Comment (tgoyne) [inline Realm/RLMRealm.mm]: ThreadLocal is definitely better here.

File: Realm/RLMRealm.mm (hunk 13 above)
Comment (tgoyne) [inline Realm/RLMRealm.mm]: I don't see any reason why it'd need to be.

File: Realm/RLMRealmUtil.mm [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+        if (fd != -1) {
+            ::close(fd);
+        }
+        fd = -1;
+    }
+
+public:
+    ~FdHolder() { close(); }
+    operator int() const { return fd; }
+    FdHolder& operator=(int newFd) {
```
Comment (alazier) [inline Realm/RLMRealmUtil.mm]: we should make sure this doens't work with FdHolder

File: Realm/RLMRealmUtil.mm [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+        int pipeFd[2];
+        ret = pipe(pipeFd);
+        if (ret == -1) {
+            return handleError(errno, error);
+        }
+
+        _shutdownReadFd = pipeFd[0];
+        _shutdownWriteFd = pipeFd[1];
+
+        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
```
Comment (alazier) [inline Realm/RLMRealmUtil.mm]: spacing looks wrong

File: Realm/RLMRealmUtil.mm [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+    return self;
+}
+
+- (void)listen {
+    // Create the runloop source
+    CFRunLoopSourceContext ctx{};
+    ctx.info = (__bridge void *)self;
+    ctx.perform = [](void *info) {
+        RLMNotifier *notifier = (__bridge RLMNotifier *)info;
+        if (RLMRealm *realm = notifier->_realm) {
```
Comment (alazier) [inline Realm/RLMRealmUtil.mm]: if statement in unecessary

File: Realm/RLMRealmUtil.mm [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+    // it should be shut down.
+    FdHolder _shutdownReadFd;
+    FdHolder _shutdownWriteFd;
+}
+
+- (instancetype)initWithRealm:(RLMRealm *)realm error:(NSError **)error {
+    self = [super init];
+    if (self) {
+        _realm = realm;
+        _runLoop = CFRunLoopGetCurrent();
```
Comment (alazier) [inline Realm/RLMRealmUtil.mm]: we should retain this.

File: Realm/RLMRealmUtil.mm [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+        CFRunLoopSourceSignal(signal);
+        // Signalling the source makes it run the next time the runloop gets
+        // to it, but doesn't make the runloop start if it's currently idle
+        // waiting for events
+        CFRunLoopWakeUp(_runLoop);
+    }
+}
+
+- (void)stop {
+    notifyFd(_shutdownWriteFd);
```
Comment (alazier) [inline Realm/RLMRealmUtil.mm]: should release runloop here

File: Realm/RLMRealmUtil.mm [hunk 19]
```diff
@@ -0,0 +1,271 @@
...
+            // Spurious wakeup; just wait again
+            continue;
+        }
+
+        // Check which file descriptor had activity: if it's the shutdown
+        // pipe, then someone called -stop; otherwise it's the named pipe
+        // and someone committed a write transaction
+        if (event.ident == (uint32_t)_shutdownReadFd) {
+            CFRunLoopSourceInvalidate(signal);
+            return;
```
Comment (alazier) [inline Realm/RLMRealmUtil.mm]: release runloop here.

File: Realm/RLMRealmUtil.mm [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+
+    // Set up the kqueue
+    // EVFILT_READ indicates that we care about data being available to read
+    // on the given file descriptor.
+    // EV_CLEAR makes it wait for the amount of data available to be read to
+    // change rather than just returning when there is any data to read.
+    struct kevent ke[2];
+    EV_SET(&ke[0], _notifyFd, EVFILT_READ, EV_ADD | EV_CLEAR, 0, 0, 0);
+    EV_SET(&ke[1], _shutdownReadFd, EVFILT_READ, EV_ADD | EV_CLEAR, 0, 0, 0);
+    kevent(_kq, ke, 2, nullptr, 0, nullptr);
```
Comment (alazier) [inline Realm/RLMRealmUtil.mm]: we should at least check the error here and assert

File: Realm/RLMRealmUtil.mm [hunk 21]
```diff
@@ -0,0 +1,271 @@
...
+            continue;
+        }
+
+        // Check which file descriptor had activity: if it's the shutdown
+        // pipe, then someone called -stop; otherwise it's the named pipe
+        // and someone committed a write transaction
+        if (event.ident == (uint32_t)_shutdownReadFd) {
+            CFRunLoopSourceInvalidate(signal);
+            return;
+        }
```
Comment (alazier) [inline Realm/RLMRealmUtil.mm]: should check that we get _notifyFd

File: Realm/Tests/InterprocessTests.m [hunk 22]
```diff
@@ -0,0 +1,329 @@
...
+        // single process do everything
+        nextRun = obj.intCol + arc4random() % 10;
+    }
+}
+
+- (void)testRecoverAfterCrash {
+    if (self.isParent) {
+        [self runChildAndWait];
+        RLMRealm *realm = RLMRealm.defaultRealm;
+        [realm beginWriteTransaction];
```
Comment (alazier) [inline Realm/Tests/InterprocessTests.m]: could actually write/validate some data to be sure.

File: Realm/Tests/InterprocessTests.m [hunk 23]
```diff
@@ -0,0 +1,329 @@
...
+    }
+    else {
+        // Tell the parent we've launched
+        [realm transactionWithBlock:^{}];
+        // Wait for a commit notification from the parent
+        [self waitForNotification:RLMRealmDidChangeNotification realm:realm block:^{}];
+    }
+}
+
+@end
```
Comment (alazier) [inline Realm/Tests/InterprocessTests.m]: newline

**Pull request merged**